Image query from root index page returns an error

### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

To reproduce the issue, clone the repo below:
https://github.com/cdaein/gatsby-image-query-issue

### Reproduction Link

https://github.com/cdaein/gatsby-image-query-issue

### Steps to Reproduce

1. clone the official tutorial example from https://github.com/gatsbyjs/tutorial-example

2. `npm i` to install packages. (ignoring peer dependency warnings below)

```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: react-server-dom-webpack@0.0.0-experimental-c8b778b7f-20220825
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"^18.2.0" from the root project
npm WARN   9 more (@mdx-js/react, gatsby, react-dom, gatsby-plugin-image, ...)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react@"0.0.0-experimental-c8b778b7f-20220825" from react-server-dom-webpack@0.0.0-experimental-c8b778b7f-20220825
npm WARN node_modules/react-server-dom-webpack
npm WARN   react-server-dom-webpack@"0.0.0-experimental-c8b778b7f-20220825" from gatsby@4.24.1
npm WARN   node_modules/gatsby
npm WARN 
npm WARN Conflicting peer dependency: react@0.0.0-experimental-c8b778b7f-20220825
npm WARN node_modules/react
npm WARN   peer react@"0.0.0-experimental-c8b778b7f-20220825" from react-server-dom-webpack@0.0.0-experimental-c8b778b7f-20220825
npm WARN   node_modules/react-server-dom-webpack
npm WARN     react-server-dom-webpack@"0.0.0-experimental-c8b778b7f-20220825" from gatsby@4.24.1
npm WARN     node_modules/gatsby
npm WARN deprecated async-cache@1.1.0: No longer maintained. Use [lru-cache](http://npm.im/lru-cache) version 7.6 or higher, and provide an asynchronous `fetchMethod` option.
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated babel-eslint@10.1.0: babel-eslint is now @babel/eslint-parser. This package will no longer receive updates.
```

3. move the `src/pages/blog/index.js` content to the root `src/pages/index.js` and make necessary changes such as import paths, component name.

4. Add `hero_image` to the root `src/pages/index.js` page query.

```
export const query = graphql`
  query {
    allMdx(sort: { fields: frontmatter___date, order: DESC }) {
      nodes {
        frontmatter {
          date(formatString: "MMMM D, YYYY")
          title
          slug
          hero_image {
            childImageSharp {
              gatsbyImageData
            }
          }
        }
        id
        excerpt
      }
    }
  }
`;
```

5. Add `<GatsbyImage image={getImage(node.frontmatter.hero_image)}/>` to the page content.

6. `gatsby clean && gatsby develop`

7. GraphQL returns an error as below:

```
 ERROR #85901  GRAPHQL

There was an error in your GraphQL query:

Cannot read properties of null (reading 'dir')

   1 | query pageUsersdaneDesktoptutorialExampleMainsrcpagesindexJs3626874412 {
   2 |   allMdx(sort: {fields: frontmatter___date, order: DESC}) {
   3 |     nodes {
   4 |       frontmatter {
   5 |         date(formatString: "MMMM D, YYYY")
   6 |         title
   7 |         slug
>  8 |         hero_image {
     |         ^
   9 |           childImageSharp {
  10 |             gatsbyImageData
  11 |           }
  12 |         }
  13 |       }
  14 |       id
  15 |       excerpt
  16 |     }
  17 |   }
  18 | }

File path: /Users/dane/Desktop/tutorial-example-main/src/pages/index.js
Url path: /
Plugin: none



  TypeError: Cannot read properties of null (reading 'dir')
  
  - resolvers.ts:535 queryNodeByPath
    [tutorial-example-main]/[gatsby]/src/schema/resolvers.ts:535:63
  
  - resolvers.ts:546 fileByPathResolver
    [tutorial-example-main]/[gatsby]/src/schema/resolvers.ts:546:14
  
  - task_queues:96 processTicksAndRejections
    node:internal/process/task_queues:96:5
  
  - async Promise.all
  
  - async Promise.all
  
  - async Promise.all
  
  - async Promise.all
```

8. Copy the exactly same image query/GatsbyImage codes back to blog `index.js` and update import path accordingly. This page query does NOT return an error. So, it must be something to do only with the root index page query. Also, as I make some changes to the code (js and mdx), sometimes, images show up on the root index page, other times not. But `gatsby clean && gatsby develop` will always return the same error on the root index page.

### Expected Result

There shouldn't be an error, and the images should show up on the root index page.

### Actual Result

GraphQL page query returns an error. And in the browser, the images do not show up.

### Environment

```shell
System:
    OS: macOS 12.6
    CPU: (10) arm64 Apple M1 Pro
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.1/bin/npm
  Browsers:
    Chrome: 106.0.5249.91
  npmPackages:
    gatsby: ^4.21.0 => 4.24.1 
    gatsby-plugin-image: ^2.20.0 => 2.24.0 
    gatsby-plugin-mdx: ^4.0.0 => 4.3.0 
    gatsby-plugin-sharp: ^4.20.0 => 4.24.0 
    gatsby-source-filesystem: ^4.20.0 => 4.24.0 
    gatsby-transformer-sharp: ^4.20.0 => 4.24.0 
  npmGlobalPackages:
    gatsby-cli: 4.24.0
```


### Config Flags

_No response_

--- a/packages/gatsby/src/schema/__tests__/fixtures/node-model.js
+++ b/packages/gatsby/src/schema/__tests__/fixtures/node-model.js
@@ -1,3 +1,6 @@
+// content of this object is not important, just the fact that it's shared in multiple nodes
+const sharedReference = {}
+
 const nodes = [
   {
     id: `person1`,
@@ -6,6 +9,7 @@ const nodes = [
     internal: { type: `Author`, contentDigest: `0` },
     name: `Person1`,
     email: `person1@example.com`,
+    sharedReference,
   },
   {
     id: `person2`,
@@ -42,6 +46,7 @@ const nodes = [
       published: false,
       date: new Date(Date.UTC(2019, 0, 1)),
     },
+    sharedReference,
   },
   {
     id: `post2`,

--- a/packages/gatsby/src/schema/__tests__/node-model.js
+++ b/packages/gatsby/src/schema/__tests__/node-model.js
@@ -502,6 +502,55 @@ describe(`NodeModel`, () => {
         const result = nodeModel.findRootNodeAncestor(obj, predicate)
         expect(result).toBe(null)
       })
+
+      describe(`multiple nodes share a reference`, () => {
+        // order of loading nodes here is important, so we check variants of loading post first and person second
+        // as well as other way around
+
+        it(`load person first, then post`, () => {
+          const sharedReference1 = nodeModel.getNodeById({
+            id: `person1`,
+          })?.sharedReference
+          const sharedReference2 = nodeModel.getNodeById({
+            id: `post1`,
+          })?.sharedReference
+
+          // Same object reference in 2 different nodes. Only `post1` has a `File` parent.
+          expect(sharedReference1 === sharedReference2).toBe(true)
+
+          const predicate = obj => obj.internal && obj.internal.type === `File`
+
+          expect(
+            nodeModel.findRootNodeAncestor(sharedReference1, predicate)?.id
+          ).toBe(`file1`)
+
+          expect(
+            nodeModel.findRootNodeAncestor(sharedReference2, predicate)?.id
+          ).toBe(`file1`)
+        })
+
+        it(`load post first, then person`, () => {
+          const sharedReference1 = nodeModel.getNodeById({
+            id: `post1`,
+          })?.sharedReference
+          const sharedReference2 = nodeModel.getNodeById({
+            id: `person1`,
+          })?.sharedReference
+
+          // Same object reference in 2 different nodes. Only `post1` has a `File` parent.
+          expect(sharedReference1 === sharedReference2).toBe(true)
+
+          const predicate = obj => obj.internal && obj.internal.type === `File`
+
+          expect(
+            nodeModel.findRootNodeAncestor(sharedReference1, predicate)?.id
+          ).toBe(`file1`)
+
+          expect(
+            nodeModel.findRootNodeAncestor(sharedReference2, predicate)?.id
+          ).toBe(`file1`)
+        })
+      })
     })
 
     describe(`createPageDependency`, () => {

--- a/packages/gatsby/src/schema/node-model.js
+++ b/packages/gatsby/src/schema/node-model.js
@@ -473,28 +473,55 @@ class LocalNodeModel {
    */
   findRootNodeAncestor(obj, predicate = null) {
     let iterations = 0
-    let node = obj
+    let ids = this._rootNodeMap.get(obj)
+    if (!ids) {
+      ids = []
+    }
+    if (obj?.parent) {
+      ids.push(obj.parent)
+    }
+    let matchingRoot = null
 
-    while (iterations++ < 100) {
-      if (predicate && predicate(node)) return node
+    console.log({ ids, obj })
 
-      const parent = getNodeById(node.parent)
-      const id = this._rootNodeMap.get(node)
-      const trackedParent = getNodeById(id)
+    if (ids) {
+      for (const id of ids) {
+        let tracked = getNodeById(id)
 
-      if (!parent && !trackedParent) {
-        const isMatchingRoot = !predicate || predicate(node)
-        return isMatchingRoot ? node : null
+        if (tracked) {
+          const visited = new Set()
+
+          while (iterations++ < 100) {
+            if (predicate && predicate(tracked)) {
+              return tracked
+            }
+
+            if (visited.has(tracked)) {
+              reporter.error(
+                `It looks like you have a node that's set its parent as itself:\n\n` +
+                  tracked
+              )
+              break
+            }
+            visited.add(tracked)
+
+            const parent = getNodeById(tracked.parent)
+
+            if (!parent) {
+              break
+            }
+
+            tracked = parent
+          }
+
+          if (tracked && !predicate) {
+            matchingRoot = tracked
+          }
+        }
       }
-
-      node = parent || trackedParent
     }
 
-    reporter.error(
-      `It looks like you have a node that's set its parent as itself:\n\n` +
-        node
-    )
-    return null
+    return matchingRoot
   }
 
   /**
@@ -877,7 +904,13 @@ const addRootNodeToInlineObject = (
 
     // don't need to track node itself
     if (!isNode) {
-      rootNodeMap.set(data, nodeId)
+      let nodeIds = rootNodeMap.get(data)
+      if (!nodeIds) {
+        nodeIds = new Set([nodeId])
+      } else {
+        nodeIds.add(nodeId)
+      }
+      rootNodeMap.set(data, nodeIds)
     }
   }
 }

--- a/packages/gatsby/src/schema/node-model.js
+++ b/packages/gatsby/src/schema/node-model.js
@@ -482,7 +482,6 @@ class LocalNodeModel {
     }
     let matchingRoot = null
 
-    console.log({ ids, obj })
 
     if (ids) {
       for (const id of ids) {

--- a/packages/gatsby/src/schema/node-model.js
+++ b/packages/gatsby/src/schema/node-model.js
@@ -482,7 +482,6 @@ class LocalNodeModel {
     }
     let matchingRoot = null
 
-
     if (ids) {
       for (const id of ids) {
         let tracked = getNodeById(id)

